README.md
# autocomplete-php package
An Atom PHP autocompleter for `autocomplete-plus`

**Now is mandatory have PHP in the enviroment PATH**

Beta version.

![Demo](https://raw.github.com/Azakur4/autocomplete-php/master/assets/img/demo.gif)

### Requirements:
- Atom `autocomplete-plus` package
  - (instll with `apm install autocomplete-plus`)
- Now is mandatory have PHP in the enviroment PATH

### Installation
- `apm install autocomplete-php`
- or install through Atom's UI

### To-Do
- [x] Get local variables from current file
- [x] Get local functions from current file (Included snippets with params)

### Maybe
- [ ] Update option for list of internal php functions

### Thanks to:
- [@amp343](https://github.com/amp343) - Add snippets to php internal functions
- [@fbereche](https://github.com/fbereche) - Help me with some ideas in the new way to get autocomplete

[![Share the love!](https://www.paypalobjects.com/pt_BR/BR/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=AMS87WQKEVEHG)


CHANGELOG.md
## 0.3.7 [Beta]
* Added a new setting to specify the path of PHP. Thanks to [@wasikuss](https://github.com/wasikuss)

## 0.3.6 [Beta]
* Added the require for path module

## 0.3.5 [Beta]
* Added the require for fs module

## 0.3.4 [Beta]
* Fix for big files with a lot of custom vars and functions

## 0.3.3 [Beta]
* Add vars completions in array

## 0.3.2 [Beta]
* More fixes when we try to get custom functions
* Better completions in differents scenarios

## 0.3.1 [Beta]
* Fix an issue with anonymous functions in php

## 0.3.0 [Beta]
* Autocomplete for custom functions with snippets for params
* Autocomplete for custom variables
* Now the process is made with php tokenizer (Now is mandatory have php bin in ur enviroment Path)

## 0.2.6 [Alpha]
* Added snippets for internal php functions by [@amp343](https://github.com/amp343)

## 0.2.5 [Alpha]
* Added constants for the provider
* Split files to separate the functions list from the keywords
* More validations for when the autocomplete have to appear

## 0.2.4 [Alpha]
* Fix for some scopes

## 0.2.3 [Alpha]
* Added php default vars ex: POST, GET, etc...

## 0.2.2 [Alpha]
* Added some snippets for keywords
* Better filter when u try to complete functions or keywords and variables

## 0.2.1 [Alpha]
* Added the autocomplete of local variables for current file (is only for testing need a lot of improvements)

## 0.2.0 - First Release [Alpha]
* Need more revision but is working


lib/test.php
<?php

abs

?>


lib/php/get_user_functions.php
<?php
    $source = '';

    while ($a = fread(STDIN, 1024)) {
        $source .= $a;
    }

    $tokens = token_get_all($source);

    $cachedFunctions = array();
    $cachedFunctionComplex = array();
    $tmpFunc = array();

    $insideFunc = false;
    $isClousure = true;
    $haveParameters = false;
    $snippetCount = 1;

    foreach($tokens as $token) {
        switch($token[0]) {
            case T_FUNCTION:
                $insideFunc = true;
                break;
            case T_STRING:
                if($insideFunc) {
                    $isClousure = false;

                    if ($token[1] !== null) {
                        if (!in_array($token[1], $cachedFunctions)) {
                            $cachedFunctions[] = $token[1];
                            $tmpFunc[] =  $token[1];
                        }
                    } else {
                        $insideFunc = false;
                    }
                }
                break;
            case '(':
                if ($insideFunc && !$isClousure) {
                    $haveParameters = true;
                }
                break;
            case ')':
                if ($insideFunc && $haveParameters) {
                    $cachedFunctionComplex[] = $tmpFunc;
                    $tmpFunc = array();
                    $snippetCount = 1;
                }

                $haveParameters = false;
                $isClousure = true;
                $insideFunc = false;
                break;
            case T_VARIABLE:
                if ($haveParameters) {
                    $tmpFunc[] = '${' . $snippetCount++ . ':' . $token[1] . '}';
                }
                break;
        }
    }


    $localFuncs = array();

    foreach ($cachedFunctionComplex as $funObj) {
        $tmp = [
            'text' => $funObj[0],
            'type' => 'function',
            'snippet' => array_shift($funObj) . '(' . implode(', ', $funObj) . ')${99}',
        ];

        array_push($localFuncs, $tmp);
    }

    echo json_encode(['user_functions' => $localFuncs]);


lib/functions/function_parameter_builder.rb
#
# for internal use only -
# please don't run me, or this
# package could be corrupted
#

module FPB

	class Base

		require 'open-uri'
		require 'nokogiri'
		require 'json'
		require 'yaml'

		def params_to_snippet(fn_name, params)
			snippets = []
			snippet_string = nil

			if params
				params.each_with_index do |param, i|
					snippets << '${'+(i+2).to_s+':'+param.to_s+'}'
				end
			end

			if !snippets.empty?
				snippet_string = fn_name.to_s+'('+snippets.join(', ')+')${'+(snippets.size+2).to_s+'}'
			else
				snippet_string = fn_name.to_s+'()${1}'
			end

			return snippet_string
		end
	end

	class YamlMap < Base

		def initialize
			@reference_path			= './phpfunc.yml'
			@source_path				= '../../functions.json'
			@destination_path		= @source_path
		end

		def get_reference_yaml
			YAML.load(File.read(@reference_path))
		end

		def get_reference_hash
			#
			# parse yaml file into a nice
			# hash structure like {name: [args], name2:..}
			#
			reference_hash = {}
			reference_data = get_reference_yaml

			reference_data['functions'].each do |fn|
				name = fn.scan(/(?:^|(?:[.!?]\s))(\w+)/).flatten.first
				params = fn.split('[').first.scan(/\$(.*?)[\s\)\,]/).flatten
				params.map!{|x| '$'+x.to_s}
				reference_hash[name] = params
			end

			return reference_hash
		end

		def get_source_hash
			JSON.parse(File.read(@source_path))['functions']
		end

		def build
			reference_data 		= get_reference_hash
			source_data				= get_source_hash
			destination_data	= {functions:source_data}

			source_data.each_with_index do |sd, i|
				reference = reference_data[sd['text']]
				destination_data[:functions][i]['snippet'] = params_to_snippet(sd['text'], reference)
			end

			File.open(@destination_path, "w") do |f|
				f.write(JSON.pretty_generate(destination_data))
			end
		end
	end

	class RemoteFetcher < Base

		def initialize
			@fn_url 							= 'http://web.archive.org/web/20150108113252/http://php.net/manual/en/function.'
			@trunk_reference_url 	= 'http://svn.php.net/viewvc/phpdoc/en/trunk/reference/'
			@source_path					= '../../functions.json'
			@destination_path			= @source_path
		end

		def build(index = nil)
			input_data 			= JSON.parse(File.read(@source_path))
			final_data 			= input_data
			final_file			= @destination_path

			if index.nil?
				index = get_starting_index(input_data['functions'])
			end

			stop = index + 50
			while index < stop
				p 'fetching: '+input_data['functions'][index]['text']+' ('+index.to_s+' / '+stop.to_s+')'
				params = fetch_required_params(input_data['functions'][index]['text'])
				final_data['functions'][index] = input_data['functions'][index]
				#final_data['functions'][index]['fetch_time'] = Time.now.to_i
				final_data['functions'][index]['snippet'] = params_to_snippet(input_data['functions'][index]['text'], params)
				index += 1
			end

			File.open(final_file, "w") do |f|
	  		f.write(JSON.pretty_generate(final_data))
			end
		end

		def get_starting_index(arr, by='fetch_time')
			#
			# start with functions that dont
			# have parameters set first
			#
			arr.each_with_index do |fn, i|
				if fn['snippet'].nil?
					return i
				end
			end
			#
			# if none are missing parameters,
			# get the index of the record with
			# minimum fetch_time (stalest)
			#
			return arr.index(arr.min_by{|x| x[by]})
		end

		def fetch_required_params(function_name)
			fn 							= function_name.gsub('_', '-')
			required_params = []

			#
			# fetch page
			#
			begin
				doc = Nokogiri::HTML(open(@fn_url+fn+".php"))
			rescue
				return required_params
			end

			#
			# access and parse definition
			#
			definition = doc.css('.methodsynopsis').to_s.split('[').first

			begin
				required_params = definition.scan(/\$(.*?)\</).flatten
			rescue
			end

			required_params.map!{|x| '$'+x.to_s }

			return required_params
		end
	end


	class Builder < Base

		attr_reader :compile_from_json_map, :compile_from_remote

		def initialize
			@yaml_map 				= FPB::YamlMap.new
			@remote_fetcher		= FPB::RemoteFetcher.new
		end

		def build_from_yaml_map
			@yaml_map.build
		end

		def build_from_remote
			@remote_fetcher.build
		end
	end
end


lib/php/get_user_all.php
<?php
    $source = '';

    while ($a = fread(STDIN, 1024)) {
        $source .= $a;
    }

    $tokens = token_get_all($source);

    $cachedVars = array();
    $cachedFunctions = array();
    $cachedFunctionComplex = array();
    $tmpFunc = array();

    $insideFunc = false;
    $isClousure = true;
    $haveParameters = false;
    $snippetCount = 1;

    foreach($tokens as $token) {
        switch($token[0]) {
            case T_FUNCTION:
                $insideFunc = true;
                break;
            case T_STRING:
                if($insideFunc) {
                    $isClousure = false;

                    if ($token[1] !== null) {
                        if (!in_array($token[1], $cachedFunctions)) {
                            $cachedFunctions[] = $token[1];
                            $tmpFunc[] =  $token[1];
                        }
                    } else {
                        $insideFunc = false;
                    }
                }
                break;
            case '(':
                if ($insideFunc && !$isClousure) {
                    $haveParameters = true;
                }
                break;
            case ')':
                if ($insideFunc && $haveParameters) {
                    $cachedFunctionComplex[] = $tmpFunc;
                    $tmpFunc = array();
                    $snippetCount = 1;
                }

                $haveParameters = false;
                $isClousure = true;
                $insideFunc = false;
                break;
            case T_VARIABLE:
                if ($haveParameters) {
                    $tmpFunc[] = '${' . $snippetCount++ . ':' . $token[1] . '}';
                }

                if (!in_array($token[1], $cachedVars)) {
                    $cachedVars[] = $token[1];
                }
                break;
        }
    }

    $localVars = array();

    foreach ($cachedVars as $var) {
        $tmp = [
            "text" => substr($var, 1),
            "type" => "variable",
        ];

        array_push($localVars, $tmp);
    }


    $localFuncs = array();

    foreach ($cachedFunctionComplex as $funObj) {
        $tmp = [
            'text' => $funObj[0],
            'type' => 'function',
            'snippet' => array_shift($funObj) . '(' . implode(', ', $funObj) . ')${99}',
        ];

        array_push($localFuncs, $tmp);
    }

    echo json_encode(['user_vars' => $localVars, 'user_functions' => $localFuncs]);
